atproto/syntax/tid.go

```diff
@@ -0,0 +1,132 @@
+package syntax
+
+import (
+	"encoding/base32"
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	Base32SortAlphabet = "234567abcdefghijklmnopqrstuvwxyz"
+)
+
+func Base32Sort() *base32.Encoding {
+	return base32.NewEncoding(Base32SortAlphabet).WithPadding(base32.NoPadding)
+}
+
+// Represents a TID in string format, as would pass Lexicon syntax validation.
+//
+// Always use [ParseTID] instead of wrapping strings directly, especially when working with network input.
+//
+// Syntax specification: https://atproto.com/specs/record-key
+type TID string
+
+func ParseTID(raw string) (TID, error) {
+	if len(raw) != 13 {
+		return "", fmt.Errorf("TID is wrong length (expected 13 chars)")
+	}
+	var tidRegex = regexp.MustCompile(`^[234567abcdefghij][234567abcdefghijklmnopqrstuvwxyz]{12}$`)
+	if !tidRegex.MatchString(raw) {
+		return "", fmt.Errorf("TID syntax didn't validate via regex")
+	}
+	return TID(raw), nil
+}
+
+// Naive (unsafe) one-off TID generation with the current time.
+//
+// You should usually use a [TIDClock] to ensure monotonic output.
+func NewTIDNow(clockId uint) TID {
+	return NewTID(time.Now().UTC().UnixMicro(), clockId)
+}
+
+func NewTIDFromInteger(v uint64) TID {
+	v = (0x7FFF_FFFF_FFFF_FFFF & v)
+	s := ""
+	for i := 0; i < 13; i++ {
+		s = string(Base32SortAlphabet[v&0x1F]) + s
+		v = v >> 5
+	}
+	return TID(s)
+}
+
+// Constructs a new TID from a UNIX timestamp (in milliseconds) and clock ID value.
+func NewTID(unixMilis int64, clockId uint) TID {
```

Comment:
I wonder if this should take a `time.Time` and extract the milliseconds for us. Thinking about the calling conventions here and Go doesn't really deal with millisecond time often.

Reply:
ah, good catch on the millis/micros, I fixed the behavior but didn't remember to update the variable names (or a test call).

I added `NewTIDFromTime`. I'd be open to renaming and have `NewTID` take a time and rename the existing to `NewTIDFromUnixMicros` (eg). I think it is fine either way.